### PR TITLE
UCT/IB: unify function to check flush rkey

### PR DESCRIPTION
**Pull request by changchengx**
## What
Unify function to check iface flush rkey.

## Why ?
make verbs & dv & devx to use one function to check iface flush rkey and related function.

### CODE REVIEW

**Review by yosefe — COMMENTED** [review]
LGTM, but shall we do it also for dc_mlx5.c?

File: src/uct/ib/rc/verbs/rc_verbs_ep.c [hunk 1]
```diff
@@ -584,17 +584,13 @@ ucs_status_t uct_rc_verbs_ep_get_address(uct_ep_h tl_ep, uct_ep_addr_t *addr)
...
-    uint8_t mr_id;
 
     rc_addr->super.flags = 0;
     uct_ib_pack_uint24(rc_addr->super.qp_num, ep->qp->qp_num);
 
-    if (md->ops->get_atomic_mr_id(md, &mr_id) == UCS_OK) {
-        ucs_assertv(uct_ib_md_is_flush_rkey_valid(md->flush_rkey),
-                    "invalid flush_rkey %x for device %s", md->flush_rkey,
-                    uct_ib_device_name(&md->dev));
+    md->ops->get_atomic_mr_id(md, &rc_addr->atomic_mr_id);
```
Comment (brminich) [inline src/uct/ib/rc/verbs/rc_verbs_ep.c:591]: looks error prone trying to fill the address without knowing whether atomic mr is supported

File: test/gtest/uct/ib/test_rc.cc [hunk 2]
```diff
@@ -193,15 +199,16 @@ UCS_TEST_P(test_rc_iface_address, size_no_flush_remote)
...
     };
     check_sizes(m_entity, sizes);
 }
 
 UCS_TEST_P(test_rc_iface_address, size_flush_remote)
 {
+    int flush_rkey_enabled = rc_iface_flush_rkey_enabled(m_entity_flush_rkey);
     map_size_t sizes = {
-        {"rc_mlx5", {10, 1}},
+        {"rc_mlx5", {flush_rkey_enabled ? 10 : 7, 1}},
```
Comment (brminich) [inline test/gtest/uct/ib/test_rc.cc:211]: why need to check here? Should not it be always enabled?

File: src/uct/ib/rc/verbs/rc_verbs_ep.c (hunk 1 above)
Comment (changchengx) [reply src/uct/ib/rc/verbs/rc_verbs_ep.c:591]: @brminich If it's not supported, it still needs to assign value(e.g. 0)
https://github.com/openucx/ucx/blob/v1.15.0-rc1/src/uct/ib/mlx5/ib_mlx5.c#L951

File: test/gtest/uct/ib/test_rc.cc (hunk 2 above)
Comment (changchengx) [reply test/gtest/uct/ib/test_rc.cc:211]: No. Think about the scenario when using [&UCT_IB_MD_OPS_NAME(dv)](https://github.com/openucx/ucx/blob/v1.15.0-rc1/src/uct/ib/base/ib_md.c#L231).

**Review by yosefe — APPROVED** [review]

File: src/uct/ib/rc/verbs/rc_verbs_ep.c (hunk 1 above)
Comment (brminich) [reply src/uct/ib/rc/verbs/rc_verbs_ep.c:591]: if it is not supported you may use smaller address and will try to access invalid memory.

File: test/gtest/uct/ib/test_rc.cc (hunk 2 above)
Comment (brminich) [reply test/gtest/uct/ib/test_rc.cc:211]: can you pls elaborate? Why wasn't the test failing before?

Comment (changchengx) [reply test/gtest/uct/ib/test_rc.cc:211]: @brminich configure with "--with-devx=no" option and run the test again.

File: src/uct/ib/rc/verbs/rc_verbs_ep.c (hunk 1 above)
Comment (changchengx) [reply src/uct/ib/rc/verbs/rc_verbs_ep.c:591]: Under this case, it won't set ```UCT_RC_VERBS_ADDR_HAS_ATOMIC_MR``` flag, right?

File: test/gtest/uct/ib/test_rc.cc (hunk 2 above)
Comment (brminich) [reply test/gtest/uct/ib/test_rc.cc:211]: ok, shall we add this config to our gtest then?

File: src/uct/ib/rc/verbs/rc_verbs_ep.c (hunk 1 above)
Comment (brminich) [reply src/uct/ib/rc/verbs/rc_verbs_ep.c:591]: now ep address length (and thus its format) is defined as below
```
iface_attr->ep_addr_len = uct_rc_iface_flush_rkey_enabled(&iface->super) ?
 sizeof(uct_rc_verbs_ep_flush_addr_t) : sizeof(uct_rc_verbs_ep_addr_t);
```
which means that we are not supposed to access `rc_addr->atomic_mr_id` if flush rkey is not enabled (e.g. outside of the if below).

File: test/gtest/uct/ib/test_rc.cc (hunk 2 above)
Comment (changchengx) [reply test/gtest/uct/ib/test_rc.cc:211]: Maybe not.
Adding it into CI test cases will make the CI pipeline too long(e.g. 1. only ```verbs```, 2. only ```verbs``` & ```dv```, 3. all ```verbs``` & ```dv``` & ```devx```)
This problem is found when I'm running some test cases to help me understand the code with using ```dv```.

File: src/uct/ib/rc/verbs/rc_verbs_ep.c (hunk 1 above)
Comment (changchengx) [reply src/uct/ib/rc/verbs/rc_verbs_ep.c:591]: @brminich Good catch.
@yosefe @brminich Let me fix this point.

Comment (changchengx) [reply src/uct/ib/rc/verbs/rc_verbs_ep.c]: @brminich It's fixed. I also checked DC, there's no similar problem.

**Comment by changchengx** [comment]
For the below failed test case, it maybe **not related with this PR**. I've verified the test case locally on CX5 & CX6DX, there's no problem.
```
2023-07-04T09:33:54.8432956Z [  FAILED  ] rc/test_ucp_tag_match_rndv_align.recv_align/1, where GetParam() = rc_v/put_zcopy,proto_v1
2023-07-04T09:33:54.8434065Z 
2023-07-04T09:33:54.8438023Z  1 FAILED TEST
```

**Comment by changchengx** [comment]
No obvious log about why the test is killed suddenly at [UCX PR(Tests new on worker 2)](https://dev.azure.com/ucfconsort/0b36e3f0-8ab9-4a48-b68b-4b2350e02c88/_apis/build/builds/65229/logs/911)
```
2023-07-05T07:48:41.1114371Z [----------] 1 test from ud_verbs/test_uct_peer_failure_multiple
2023-07-05T07:48:41.1114757Z [ RUN      ] ud_verbs/test_uct_peer_failure_multiple.test/1 <ud_verbs/mlx5_2:1>
2023-07-05T07:48:56.0400370Z make: *** [test] Killed
2023-07-05T07:48:56.1719432Z make: Leaving directory `/scrap/azure/agent-09/AZP_WORKSPACE/2/s/build-test/test/gtest'
2023-07-05T07:48:57.1091135Z ##[error]Bash exited with code '2'.
2023-07-05T07:48:57.6609479Z ##[section]Finishing: Run ./contrib/test_jenkins.sh
2023-07-05T07:48:57.7140707Z ##[section]Starting: Checkout openucx/ucx@refs/pull/9147/merge to s
```

**Comment by changchengx** [comment]
Below error log maybe not related with this PR in [UCX PR(Tests new on worker 2)](https://dev.azure.com/ucfconsort/0b36e3f0-8ab9-4a48-b68b-4b2350e02c88/_apis/build/builds/65229/logs/938)
```
023-07-05T20:07:11.2404619Z [ RUN      ] shm_ib/test_ucp_sockaddr.concurrent_disconnect_c2s/2 <shm,ib/tag,sa_data_v2>
2023-07-05T20:07:12.1547921Z ==8703== 
2023-07-05T20:07:12.1548802Z ==8703== Process terminating with default action of signal 15 (SIGTERM)
2023-07-05T20:07:12.1550076Z ==8703==    at 0x769D8D7: ioctl (in /usr/lib64/libc-2.17.so)
2023-07-05T20:07:12.1551157Z ==8703==    by 0x5EA094E: execute_ioctl (in /usr/lib64/libibverbs.so.1.14.40.0)
2023-07-05T20:07:12.1551692Z ==8703==    by 0x5E9FCBA: ??? (in /usr/lib64/libibverbs.so.1.14.40.0)
2023-07-05T20:07:12.1552163Z ==8703==    by 0x5E9E1EF: ibv_cmd_query_port (in /usr/lib64/libibverbs.so.1.14.40.0)
2023-07-05T20:07:12.1552611Z ==8703==    by 0x6107C15: ??? (in /usr/lib64/libmlx5.so.1.23.40.0)
2023-07-05T20:07:12.1553048Z ==8703==    by 0x60F22B1: ??? (in /usr/lib64/libmlx5.so.1.23.40.0)
2023-07-05T20:07:12.1553557Z ==8703==    by 0x60F2B0A: ??? (in /usr/lib64/libmlx5.so.1.23.40.0)
2023-07-05T20:07:12.1554016Z ==8703==    by 0x5EA54D2: verbs_open_device (in /usr/lib64/libibverbs.so.1.14.40.0)
```

**Comment by changchengx** [comment]
Below failure maybe not related with this PR in [UCX PR(Tests new on worker 0)]()
```
2023-07-05T15:35:57.4025484Z [ RUN      ] tcp/test_ucp_tag_match_rndv.exp_huge_mix/1 <tcp/auto,proto_v1>
2023-07-05T15:36:01.3905061Z unknown file: Failure
2023-07-05T15:36:01.3905648Z C++ exception with description "std::bad_alloc" thrown in the test body.
2023-07-05T15:36:01.3905990Z [  FAILED  ] tcp/test_ucp_tag_match_rndv.exp_huge_mix/1, where GetParam() = tcp/auto,proto_v1 (3827 ms)


2023-07-05T16:11:46.2699930Z [  FAILED  ] 1 test, listed below:
2023-07-05T16:11:46.2702052Z [  FAILED  ] tcp/test_ucp_tag_match_rndv.exp_huge_mix/1, where GetParam() = tcp/auto,proto_v1
2023-07-05T16:11:46.2702864Z 
2023-07-05T16:11:46.2703378Z  1 FAILED TEST
```

**Comment by changchengx** [comment]
The failed test cases report below log
```
##[error]The job running on agent swx-rdmz-ucx-new-01-02 ran longer than the maximum time of 360 minutes. For more information, see https://go.microsoft.com/fwlink/?linkid=2077134
```

**Comment by changchengx** [comment]
Let me solve the code conflict first.

**Comment by changchengx** [comment]
Solved all conflicts & Passed all CI test cases.

**Review by yosefe — APPROVED** [review]
@changchengx can you pls squash?

**Review by yosefe — APPROVED** [review]

**Comment by changchengx** [comment]
The failure maybe not related with this PR in [UCX PR(Tests new on worker0)](https://dev.azure.com/ucfconsort/0b36e3f0-8ab9-4a48-b68b-4b2350e02c88/_apis/build/builds/65547/logs/944)
```
2023-07-10T12:42:57.9815201Z [       OK ] ud/test_ucp_sockaddr_conn_request.conn_request_query_worker_id/0 (2200 ms)
2023-07-10T12:42:57.9815797Z [ RUN      ] ud/test_ucp_sockaddr_conn_request.conn_request_query_worker_id/4 <ud_v/tag,not_all_devs>
2023-07-10T12:42:57.9816339Z ==11952== 
2023-07-10T12:42:57.9816773Z ==11952== Process terminating with default action of signal 15 (SIGTERM)
2023-07-10T12:42:57.9817529Z ==11952==    at 0x674CD12: pthread_cond_timedwait@@GLIBC_2.3.2 (in /usr/lib64/libpthread-2.17.so)
2023-07-10T12:42:57.9818349Z ==11952==    by 0x5F7FD5: ucs::watchdog_func(void*) (test_helpers.cc:50)
2023-07-10T12:42:57.9818959Z ==11952==    by 0x6748DD4: start_thread (in /usr/lib64/libpthread-2.17.so)
2023-07-10T12:42:57.9819544Z ==11952==    by 0x76A7EAC: clone (in /usr/lib64/libc-2.17.so)
```

**Pull request merged**